### PR TITLE
VTHP-1870: Resolve nats no longer consuming messages because of being…

### PR DESCRIPTION
**Pull request by Amnatehreem**

### CODE REVIEW

**Comment by waeljammal** [comment]
# go-cover-view


<details> <summary> github.com/azarc-io/vth-faas-sdk-go/pkg/spark/v1/plugin.go </summary>

```
    1: package sparkv1
    2:
    3: import (
    4: 	"context"
    5: 	"errors"
    6: 	"github.com/cenkalti/backoff/v4"
    7: 	"github.com/google/uuid"
    8: 	"github.com/hashicorp/go-plugin"
    9: 	"github.com/nats-io/nats.go"
   10: 	"github.com/nats-io/nats.go/jetstream"
   11: 	"github.com/rs/zerolog/log"
   12: 	"time"
   13: )
   14:
   15: var (
   16: 	ErrInvalidStageResultMimeType = errors.New("stage result expects mime-type of application/json")
   17: 	ErrTemporalIoNotSupported     = errors.New("temporal IO provider does not support input/output referencing")
   18: )
   19:
   20: /************************************************************************/
   21: // TYPES
   22: /************************************************************************/
   23:
   24: type sparkPlugin struct {
   25: 	config *Config
   26: 	chain  *SparkChain
   27: 	ctx    context.Context
   28: 	nc     *nats.Conn
   29: }
   30:
   31: /************************************************************************/
   32: // SERVER
   33: /************************************************************************/
   34:
X  35: func newSparkPlugin(ctx context.Context, cfg *Config, chain *SparkChain) *sparkPlugin {
X  36: 	return &sparkPlugin{ctx: ctx, config: cfg, chain: chain}
X  37: }
   38:
X  39: func (s *sparkPlugin) start() error {
X  40: 	nc, err := s.createNatsClient()
X  41: 	if err != nil {
X  42: 		return err
X  43: 	}
X  44: 	s.nc = nc
X  45:
X  46: 	js, err := jetstream.New(nc)
X  47: 	if err != nil {
X  48: 		return err
X  49: 	}
X  50: 	store, err := js.ObjectStore(s.ctx, s.config.NatsBucket)
X  51: 	if err != nil {
X  52: 		return err
X  53: 	}
   54:
X  55: 	wf, err := NewJobWorkflow(s.ctx, uuid.NewString(), s.chain,
X  56: 		WithConfig(s.config), WithNatsClient(nc), WithObjectStore(store))
X  57: 	if err != nil {
X  58: 		return err
X  59: 	}
   60:
X  61: 	err = s.createEventConsumer(js, wf)
X  62: 	if err != nil {
X  63: 		return err
X  64: 	}
   65:
X  66: 	plugin.Serve(&plugin.ServeConfig{
X  67: 		HandshakeConfig: plugin.HandshakeConfig{
X  68: 			ProtocolVersion:  1,
X  69: 			MagicCookieKey:   "BASIC_PLUGIN",
X  70: 			MagicCookieValue: s.config.Id,
X  71: 		},
X  72: 		Plugins: make(map[string]plugin.Plugin),
X  73: 	})
X  74:
X  75: 	return nil
   76: }
   77:
X  78: func (s *sparkPlugin) createEventConsumer(js jetstream.JetStream, wf JobWorkflow) error {
X  79: 	stream, err := js.CreateOrUpdateStream(s.ctx, jetstream.StreamConfig{
X  80: 		Name: s.config.NatsRequestStreamName,
X  81: 	})
X  82: 	if err != nil {
X  83: 		return err
X  84: 	}
   85:
X  86: 	consumer, err := stream.CreateOrUpdateConsumer(context.Background(), jetstream.ConsumerConfig{
X  87: 		Name:              s.config.Id,
X  88: 		FilterSubject:     s.config.NatsRequestSubject,
X  89: 		AckPolicy:         jetstream.AckExplicitPolicy,
X  90: 		AckWait:           s.config.Timeout,
X  91: 		MaxDeliver:        1,
X  92: 		MaxAckPending:     1,
X  93: 		InactiveThreshold: time.Hour,
X  94: 	})
X  95: 	if err != nil {
X  96: 		log.Error().Err(err).Msgf("could not create consumer for subject %s", s.config.NatsRequestSubject)
X  97: 		return err
X  98: 	}
   99:
X 100: 	go func() {
X 101: 		var lastConsumedTime = time.Now()
X 102: 	loop:
X 103: 		for {
X 104: 			select {
  105: 			// on consumer stopped
X 106: 			case <-s.ctx.Done():
X 107: 				log.Info().Msgf("stopping consumer")
X 108: 				return
X 109: 			default:
X 110: 				batch, err := consumer.Fetch(15, jetstream.FetchMaxWait(time.Second*15))
X 111: 				if err != nil {
X 112: 					log.Error().Err(err).Msgf("failed to fetch job request messages, will retry shortly")
X 113: 					continue
  114: 				}
  115:
X 116: 				var received bool
X 117: 				for msg := range batch.Messages() {
X 118: 					received = true
X 119: 					go func(m jetstream.Msg) {
X 120: 						m.Ack()
X 121: 						wf.Run(m)
X 122: 					}(msg)
  123: 				}
  124:
X 125: 				if received {
X 126: 					lastConsumedTime = time.Now()
X 127: 				} else if time.Since(lastConsumedTime) > time.Minute*45 {
X 128: 					err := backoff.Retry(func() error {
X 129: 						return s.createEventConsumer(js, wf)
X 130: 					}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 3))
  131:
X 132: 					if err != nil {
X 133: 						log.Error().Err(err).Msgf(
X 134: 							"failed to re-subscribe audit event consumer after it became idle for too long")
X 135: 					}
  136:
X 137: 					break loop
  138: 				}
  139: 			}
  140: 		}
  141: 	}()
  142:
X 143: 	return nil
  144: }
  145:
X 146: func (s *sparkPlugin) stop() {
X 147: 	if s.nc != nil {
X 148: 		_ = s.nc.Drain()
X 149: 	}
  150: }
  151:
X 152: func (s *sparkPlugin) createNatsClient() (*nats.Conn, error) {
X 153: 	return nats.Connect(s.config.Nats.Address)
X 154: }

```

</details>

File: pkg/spark/v1/plugin.go [hunk 1]
```diff
@@ -89,25 +113,33 @@ func (s *sparkPlugin) start() error {
...
+					received = true
 					go func(m jetstream.Msg) {
 						m.Ack()
 						wf.Run(m)
 					}(msg)
 				}
+
+				if received {
+					lastConsumedTime = time.Now()
+				} else if time.Since(lastConsumedTime) > time.Minute*45 {
```
Comment (azarc-jono-langley) [inline pkg/spark/v1/plugin.go:127]: this should be a const `time.Minute*45`

File: pkg/spark/v1/plugin.go [hunk 2]
```diff
@@ -89,25 +113,33 @@ func (s *sparkPlugin) start() error {
...
+					err := backoff.Retry(func() error {
+						return s.createEventConsumer(js, wf)
+					}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 3))
```
Comment (azarc-jono-langley) [inline pkg/spark/v1/plugin.go:130]: 3 should be a const.

**Comment by waeljammal** [comment]
# go-cover-view


<details> <summary> github.com/azarc-io/vth-faas-sdk-go/pkg/spark/v1/plugin.go </summary>

```
    1: package sparkv1
    2:
    3: import (
    4: 	"context"
    5: 	"errors"
    6: 	"github.com/cenkalti/backoff/v4"
    7: 	"github.com/google/uuid"
    8: 	"github.com/hashicorp/go-plugin"
    9: 	"github.com/nats-io/nats.go"
   10: 	"github.com/nats-io/nats.go/jetstream"
   11: 	"github.com/rs/zerolog/log"
   12: 	"time"
   13: )
   14:
   15: var (
   16: 	ErrInvalidStageResultMimeType = errors.New("stage result expects mime-type of application/json")
   17: 	ErrTemporalIoNotSupported     = errors.New("temporal IO provider does not support input/output referencing")
   18: )
   19:
   20: /************************************************************************/
   21: // TYPES
   22: /************************************************************************/
   23:
   24: type sparkPlugin struct {
   25: 	config *Config
   26: 	chain  *SparkChain
   27: 	ctx    context.Context
   28: 	nc     *nats.Conn
   29: }
   30:
   31: /************************************************************************/
   32: // SERVER
   33: /************************************************************************/
   34:
X  35: func newSparkPlugin(ctx context.Context, cfg *Config, chain *SparkChain) *sparkPlugin {
X  36: 	return &sparkPlugin{ctx: ctx, config: cfg, chain: chain}
X  37: }
   38:
X  39: func (s *sparkPlugin) start() error {
X  40: 	nc, err := s.createNatsClient()
X  41: 	if err != nil {
X  42: 		return err
X  43: 	}
X  44: 	s.nc = nc
X  45:
X  46: 	js, err := jetstream.New(nc)
X  47: 	if err != nil {
X  48: 		return err
X  49: 	}
X  50: 	store, err := js.ObjectStore(s.ctx, s.config.NatsBucket)
X  51: 	if err != nil {
X  52: 		return err
X  53: 	}
   54:
X  55: 	wf, err := NewJobWorkflow(s.ctx, uuid.NewString(), s.chain,
X  56: 		WithConfig(s.config), WithNatsClient(nc), WithObjectStore(store))
X  57: 	if err != nil {
X  58: 		return err
X  59: 	}
   60:
X  61: 	err = s.createEventConsumer(js, wf)
X  62: 	if err != nil {
X  63: 		return err
X  64: 	}
   65:
X  66: 	plugin.Serve(&plugin.ServeConfig{
X  67: 		HandshakeConfig: plugin.HandshakeConfig{
X  68: 			ProtocolVersion:  1,
X  69: 			MagicCookieKey:   "BASIC_PLUGIN",
X  70: 			MagicCookieValue: s.config.Id,
X  71: 		},
X  72: 		Plugins: make(map[string]plugin.Plugin),
X  73: 	})
X  74:
X  75: 	return nil
   76: }
   77:
X  78: func (s *sparkPlugin) createEventConsumer(js jetstream.JetStream, wf JobWorkflow) error {
X  79: 	stream, err := js.CreateOrUpdateStream(s.ctx, jetstream.StreamConfig{
X  80: 		Name: s.config.NatsRequestStreamName,
X  81: 	})
X  82: 	if err != nil {
X  83: 		return err
X  84: 	}
   85:
X  86: 	consumer, err := stream.CreateOrUpdateConsumer(context.Background(), jetstream.ConsumerConfig{
X  87: 		Name:              s.config.Id,
X  88: 		FilterSubject:     s.config.NatsRequestSubject,
X  89: 		AckPolicy:         jetstream.AckExplicitPolicy,
X  90: 		AckWait:           s.config.Timeout,
X  91: 		MaxDeliver:        1,
X  92: 		MaxAckPending:     1,
X  93: 		InactiveThreshold: maxInactiveConsumerDuration,
X  94: 	})
X  95: 	if err != nil {
X  96: 		log.Error().Err(err).Msgf("could not create consumer for subject %s", s.config.NatsRequestSubject)
X  97: 		return err
X  98: 	}
   99:
X 100: 	go func() {
X 101: 		var lastConsumedTime = time.Now()
X 102: 	loop:
X 103: 		for {
X 104: 			select {
  105: 			// on consumer stopped
X 106: 			case <-s.ctx.Done():
X 107: 				log.Info().Msgf("stopping consumer")
X 108: 				return
X 109: 			default:
X 110: 				batch, err := consumer.Fetch(15, jetstream.FetchMaxWait(maxConsumerFetchWait))
X 111: 				if err != nil {
X 112: 					log.Error().Err(err).Msgf("failed to fetch job request messages, will retry shortly")
X 113: 					continue
  114: 				}
  115:
X 116: 				var received bool
X 117: 				for msg := range batch.Messages() {
X 118: 					received = true
X 119: 					go func(m jetstream.Msg) {
X 120: 						m.Ack()
X 121: 						wf.Run(m)
X 122: 					}(msg)
  123: 				}
  124:
X 125: 				if received {
X 126: 					lastConsumedTime = time.Now()
X 127: 				} else if time.Since(lastConsumedTime) > maxInactiveResetConsumerDuration {
X 128: 					err := backoff.Retry(func() error {
X 129: 						return s.createEventConsumer(js, wf)
X 130: 					}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 3))
  131:
X 132: 					if err != nil {
X 133: 						log.Error().Err(err).Msgf(
X 134: 							"failed to re-subscribe audit event consumer after it became idle for too long")
X 135: 					}
  136:
X 137: 					break loop
  138: 				}
  139: 			}
  140: 		}
  141: 	}()
  142:
X 143: 	return nil
  144: }
  145:
X 146: func (s *sparkPlugin) stop() {
X 147: 	if s.nc != nil {
X 148: 		_ = s.nc.Drain()
X 149: 	}
  150: }
  151:
X 152: func (s *sparkPlugin) createNatsClient() (*nats.Conn, error) {
X 153: 	return nats.Connect(s.config.Nats.Address)
X 154: }

```

</details>

**Review by azarc-jono-langley — APPROVED** [review]

**Comment by waeljammal** [comment]
# go-cover-view


<details> <summary> github.com/azarc-io/vth-faas-sdk-go/pkg/spark/v1/plugin.go </summary>

```
    1: package sparkv1
    2:
    3: import (
    4: 	"context"
    5: 	"errors"
    6: 	"github.com/cenkalti/backoff/v4"
    7: 	"github.com/google/uuid"
    8: 	"github.com/hashicorp/go-plugin"
    9: 	"github.com/nats-io/nats.go"
   10: 	"github.com/nats-io/nats.go/jetstream"
   11: 	"github.com/rs/zerolog/log"
   12: 	"time"
   13: )
   14:
   15: var (
   16: 	ErrInvalidStageResultMimeType = errors.New("stage result expects mime-type of application/json")
   17: 	ErrTemporalIoNotSupported     = errors.New("temporal IO provider does not support input/output referencing")
   18: )
   19:
   20: /************************************************************************/
   21: // TYPES
   22: /************************************************************************/
   23:
   24: type sparkPlugin struct {
   25: 	config *Config
   26: 	chain  *SparkChain
   27: 	ctx    context.Context
   28: 	nc     *nats.Conn
   29: }
   30:
   31: /************************************************************************/
   32: // SERVER
   33: /************************************************************************/
   34:
X  35: func newSparkPlugin(ctx context.Context, cfg *Config, chain *SparkChain) *sparkPlugin {
X  36: 	return &sparkPlugin{ctx: ctx, config: cfg, chain: chain}
X  37: }
   38:
X  39: func (s *sparkPlugin) start() error {
X  40: 	nc, err := s.createNatsClient()
X  41: 	if err != nil {
X  42: 		return err
X  43: 	}
X  44: 	s.nc = nc
X  45:
X  46: 	js, err := jetstream.New(nc)
X  47: 	if err != nil {
X  48: 		return err
X  49: 	}
X  50: 	store, err := js.ObjectStore(s.ctx, s.config.NatsBucket)
X  51: 	if err != nil {
X  52: 		return err
X  53: 	}
   54:
X  55: 	wf, err := NewJobWorkflow(s.ctx, uuid.NewString(), s.chain,
X  56: 		WithConfig(s.config), WithNatsClient(nc), WithObjectStore(store))
X  57: 	if err != nil {
X  58: 		return err
X  59: 	}
   60:
X  61: 	err = s.createEventConsumer(js, wf)
X  62: 	if err != nil {
X  63: 		return err
X  64: 	}
   65:
X  66: 	plugin.Serve(&plugin.ServeConfig{
X  67: 		HandshakeConfig: plugin.HandshakeConfig{
X  68: 			ProtocolVersion:  1,
X  69: 			MagicCookieKey:   "BASIC_PLUGIN",
X  70: 			MagicCookieValue: s.config.Id,
X  71: 		},
X  72: 		Plugins: make(map[string]plugin.Plugin),
X  73: 	})
X  74:
X  75: 	return nil
   76: }
   77:
X  78: func (s *sparkPlugin) createEventConsumer(js jetstream.JetStream, wf JobWorkflow) error {
X  79: 	stream, err := js.CreateOrUpdateStream(s.ctx, jetstream.StreamConfig{
X  80: 		Name: s.config.NatsRequestStreamName,
X  81: 	})
X  82: 	if err != nil {
X  83: 		return err
X  84: 	}
   85:
X  86: 	consumer, err := stream.CreateOrUpdateConsumer(context.Background(), jetstream.ConsumerConfig{
X  87: 		Name:              s.config.Id,
X  88: 		FilterSubject:     s.config.NatsRequestSubject,
X  89: 		AckPolicy:         jetstream.AckExplicitPolicy,
X  90: 		AckWait:           s.config.Timeout,
X  91: 		MaxDeliver:        maxConsumerDeliver,
X  92: 		MaxAckPending:     maxConsumerAckPending,
X  93: 		InactiveThreshold: maxInactiveConsumerDuration,
X  94: 	})
X  95: 	if err != nil {
X  96: 		log.Error().Err(err).Msgf("could not create consumer for subject %s", s.config.NatsRequestSubject)
X  97: 		return err
X  98: 	}
   99:
X 100: 	go func() {
X 101: 		var lastConsumedTime = time.Now()
X 102: 	loop:
X 103: 		for {
X 104: 			select {
  105: 			// on consumer stopped
X 106: 			case <-s.ctx.Done():
X 107: 				log.Info().Msgf("stopping consumer")
X 108: 				return
X 109: 			default:
X 110: 				batch, err := consumer.Fetch(ConsumerBatch, jetstream.FetchMaxWait(maxConsumerFetchWait))
X 111: 				if err != nil {
X 112: 					log.Error().Err(err).Msgf("failed to fetch job request messages, will retry shortly")
X 113: 					continue
  114: 				}
  115:
X 116: 				var received bool
X 117: 				for msg := range batch.Messages() {
X 118: 					received = true
X 119: 					go func(m jetstream.Msg) {
X 120: 						m.Ack()
X 121: 						wf.Run(m)
X 122: 					}(msg)
  123: 				}
  124:
X 125: 				if received {
X 126: 					lastConsumedTime = time.Now()
X 127: 				} else if time.Since(lastConsumedTime) > maxInactiveResetConsumerDuration {
X 128: 					err := backoff.Retry(func() error {
X 129: 						return s.createEventConsumer(js, wf)
X 130: 					}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxConsumerCreationRetries))
  131:
X 132: 					if err != nil {
X 133: 						log.Error().Err(err).Msgf(
X 134: 							"failed to re-subscribe audit event consumer after it became idle for too long")
X 135: 					}
  136:
X 137: 					break loop
  138: 				}
  139: 			}
  140: 		}
  141: 	}()
  142:
X 143: 	return nil
  144: }
  145:
X 146: func (s *sparkPlugin) stop() {
X 147: 	if s.nc != nil {
X 148: 		_ = s.nc.Drain()
X 149: 	}
  150: }
  151:
X 152: func (s *sparkPlugin) createNatsClient() (*nats.Conn, error) {
X 153: 	return nats.Connect(s.config.Nats.Address)
X 154: }

```

</details>

**Pull request merged**